divintseg/_core.py

```diff
@@ -422,3 +422,151 @@ def similarity(
+    +----------+-------------+-----------------------+
+    | Region 2 | Subregion E | 10 / (10 + 90) = 0.1  |
+    +----------+-------------+-----------------------+
+
+    Next, we can compute the fraction of all S's in
+    each subregion of each region. There are 150 S's
+    in Region 1 and 10 S's in region 2, therefore, we have:
+
+    +----------+-------------+--------------------------------+
+    | Region   | Subregion   | Fraction of all As in Region   |
+    +==========+=============+================================+
+    | Region 1 | Subregion A | 100 / 150 = 0.6667             |
+    +----------+-------------+--------------------------------+
+    | Region 1 | Subregion B | 50 / 150 = 0.3333              |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion C | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion D | 0 / 10 = 0.0000                |
+    +----------+-------------+--------------------------------+
+    | Region 2 | Subregion E | 10 / 10 = 1.0000               |
+    +----------+-------------+--------------------------------+
+
+    Finally, for each subregion, we multiply these together and
+    add them up the values for the subregions in each region.
+    For Region 1, we get
+
+    .. math::
+        (0.6667 * 1.0) + (0.3333 * 0.5) = 0.8333.
+
+    For Region 2, we get
+
+    .. math::
+        0.0 * 0.0 + 0.0 * 0.0 + 1.000 * 0.1 = 0.1.
+
+    Note that the implentation may not do this math exactly as
+    specified here, but it will do something equivalent.
+    """
+    if by is None:
+        df_communities["REGION"] = "Region " + df_communities.index.astype(str)
+        by = "REGION"
+    if over is None:
+        df_communities["SUBREGION"] = "Subregion " + df_communities.index.astype(str)
+        over = "SUBREGION"
+    else:
+        subregion_groups = df_communities.groupby(over)
+        for _, group in subregion_groups:
+            if len(group[by].unique()) > 1:
+                raise ValueError("Subregions must have the same region")
+                # If 2 subregions are the same, their region must be the same
+    df_by = df_communities.groupby(by).sum(numeric_only=True)[group_name]
+    df_over = df_communities.groupby(over).sum(numeric_only=True)
+    likelihood = df_over[group_name] / df_over.sum(axis="columns")
+
+    def get_region_pop(row):
+        # Used with apply to add region and region population columns
+        row_num = df_communities.loc[df_communities[over] == row.name]
+        region = row_num[by].values[0]
+        row[by] = region
+        row["Region Population"] = df_by[region]
+        return row
+    df_over = df_over.apply(get_region_pop, axis=1)
+    frac_in_region = df_over[group_name]/df_over["Region Population"]
+    df_over["Product"] = likelihood*frac_in_region
+    df_over = df_over.groupby(by).sum(numeric_only=True).reset_index()
```

Comment:
I think `"Product"` is the only column you care about, so you could do something like `df_over.groupby(by)['Product'].sum().reset_index(drop=True)` to get just the sum of the products in each group.